### PR TITLE
Initial clean up of System.Threading.Tasks.Channels

### PR DESCRIPTION
**Pull request by stephentoub**
In support of https://github.com/dotnet/corefxlab/issues/1731, some clean up work on Channels, including removing experimental API surface area (e.g. CaseBuilder, IAsyncEnumerator, etc.), which can be added back later when the support is ready, and removing from the public surface area internal dependencies that would be better off in a lower-level library (though some of the public surface area still there should likely end up in System.Threading.Tasks.Extensions, or something like that).  Also includes some perf work, but there's more of that to be done separately.

Fixes https://github.com/dotnet/corefxlab/issues/1127
Fixes https://github.com/dotnet/corefxlab/issues/1723
Contributes to https://github.com/dotnet/corefxlab/issues/1731

### CODE REVIEW

File: src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/WritableChannel.cs [hunk 1]
```diff
@@ -18,9 +18,9 @@ public abstract class WritableChannel<T>
         /// <param name="error">An <see cref="Exception"/> indicating the failure causing no more data to be written, or null for success.</param>
         /// <returns>
         /// true if this operation successfully completes the channel; otherwise, false if the channel could not be marked for completion,
-        /// for example due to having already been marked as such.
+        /// for example due to having already been marked as such, or due to not supporting completion.
         /// </returns>
-        public abstract bool TryComplete(Exception error = null);
+        public virtual bool TryComplete(Exception error = null) => false;
```
Comment (KrzysztofCwalina) [inline src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/WritableChannel.cs]: Why this change?

Comment (stephentoub) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/WritableChannel.cs]: It's essentially saying that the default is an infinite channel; Completion returns a never-completing task by default, and TryComplete returns false to say you can't complete it.  If a derived channel wants to override it, it of course can, and do so in a way that appropriately coordinates between the read side and the write side.  We could go back to having them both be abstract, but it seemed like an unnecessary additional thing to be forced to override in order to implement a custom channel.

I don't feel strongly about it, though.  Happy to change it back if you disagree.

Comment (KrzysztofCwalina) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/WritableChannel.cs]: Ah, thanks. The change makes sense.

File: src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs [hunk 2]
```diff
@@ -192,52 +223,5 @@ public void Dispose()
                 }
             }
         }
-
-        /// <summary>Gets an async enumerator of the data in this channel.</summary>
-        /// <param name="cancellationToken">The cancellation token to use to cancel the asynchronous enumeration.</param>
-        /// <returns>The async enumerator.</returns>
-        public virtual IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default) =>
```
Comment (BrennanConroy) [inline src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs]: @davidfowl @anurse We use this for streaming in SignalR. Looks easy enough to make into an extension method for our internal use until a public API is added back.

@stephentoub Is there an ETA for this support, like 2.1?

Comment (stephentoub) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs]: I removed it because I don't want it to conflict with the real `IAsyncEnumerable<T>/IAsyncEnumerator<T>` we intend to deliver along with C# 8.  I do not expect this to be in 2.1.
https://github.com/dotnet/csharplang/blob/master/proposals/async-streams.md

Comment (anurse) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs]: Ok, I think we'll probably just copy this extension method into our code base as internal code (and plan to replace it with the real thing when that comes out).

Comment (stephentoub) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs]: Sounds good.  Thanks.

Comment (davidfowl) [reply src/System.Threading.Tasks.Channels/System/Threading/Tasks/Channels/ReadableChannel.cs]: ZOMG 😄
